### PR TITLE
Fix/issue 1157

### PR DESCRIPTION
**Pull request by perwestling**
Presentation has now been improved:
Screenshot from 2020-07-24 13-02-21

Made some changes in the ability presentation:
- Do not show abilities without name
- Make presentation layout more dynamic
The reason for having abilities without name is to hide them from presentation. In some places an ability can be represented as multiple attributes internally.

Also, it will also mean that adding abilities to appear will be an intentional action.

The dynamic part of presentation was created by @yzemaze

### CODE REVIEW

File: lib/engine/config/game/g_1846.rb [hunk 1]
```diff
@@ -491,6 +495,8 @@ module G1846
       "abilities": [
         {
           "type": "token",
+          "name": "Reserved Eerie token",
```
Comment (jenf) [inline lib/engine/config/game/g_1846.rb:498]: Erie?

File: lib/engine/config/game/g_1846.rb [hunk 2]
```diff
@@ -444,6 +446,8 @@ module G1846
       "abilities": [
         {
           "type": "token",
+          "name": "Reserved Cincinatti token",
```
Comment (jenf) [inline lib/engine/config/game/g_1846.rb:449]: Cincinnati

**Comment by jenf** [comment]
Apart from the typos looks good for me, will need Toby to review.

**Comment by dfannius** [comment]
I would write "Ability" instead of "Attribute", and "Cost" instead of "Value".

**Comment by perwestling** [comment]
> I would write "Ability" instead of "Attribute", and "Cost" instead of "Value".

@dfannius It is not always Cost. E.g. in 18AL uses this for its Historical objective, and then the "display_value" is the name of the city that is the destination. Also in 18AL when Coal Field token is assigned to one of the 5 cities, the hex assigned is shown in Value. 
The third one in 18AL is the route bonuses, where the name of the route is shown in Attribute and the bonus ($20 or $40) is shown in Value.

So I saw it as a value to the attribute....

For 1846 one could consider to instead use e.g. PRR
Attribute                                Value
Reserved $40 token          Cincinatti

The ones I was least happy with is the "Free tile lay       $0", I think it would be better to show the 5 hexes in the Value.

Ability/Attribute - yes, I think that can be a good one.

Cost/Value - do we have something more neutral?

**Comment by perwestling** [comment]
About 18AL historical objective vs Coal Field token: Maybe I should be consistent and use hex name in both cases?  So G4 instead of Birmingham. (Alternatively to use city name also for Coal Field token.)

**Comment by dfannius** [comment]
I see, your "value" means the value of a variable, not a monetary value (which is what people would interpret the word as).

If the cells in the second column can mean very different things from row to row, I would not even make a header for it. Ideally the cells could say things like "costs $40" but I know there is not much space. Parenthetical expressions like "($40)" are used in accounting to denote negative numbers and maybe could get the idea of a cost across.

**Comment by perwestling** [comment]
It is the associated value of the ability/attribute, and if it is a cost, a bonus, an income, a hex, a city depends on the context of the ability....  
Removing the column header for that part is an idea, but I think it can be confusing if the Company table above Ability/Attribute shows Value and Income on the right part.

Another variant is to just use "name" and then each game can put whatever they like in that name, and not have anything to the right.
So in 18AL it would be
`Historical objective Birmingham`
And in 1846 is would be
`Reserved $40/$60 Ft. Wayne token`

**Comment by perwestling** [comment]
18AL version:
Screenshot from 2020-07-24 16-05-21

1846 updated version:
Screenshot from 2020-07-24 16-19-31

File: assets/app/view/game/corporation.rb [hunk 3]
```diff
@@ -48,8 +48,10 @@ def render
           children << h(Companies, owner: @corporation, game: @game) if @corporation.companies.any?
         end
 
-        abilities = @corporation.all_abilities.select { |ability| ability.owner.corporation? }
-        children << render_abilities(abilities) if abilities&.any?
+        abilities_to_display = @corporation.all_abilities.select do |ability|
+          ability.owner.corporation? && !ability.name.nil?
```
Comment (tobymao) [inline assets/app/view/game/corporation.rb:52]: ability.name

File: assets/app/view/game/corporation.rb [hunk 4]
```diff
@@ -48,8 +48,10 @@ def render
...
+        end
+        children << render_abilities(abilities_to_display) if abilities_to_display&.any?
```
Comment (tobymao) [inline assets/app/view/game/corporation.rb:54]: abilities_to_display will never be nil no need for &

File: assets/app/view/game/corporation.rb [hunk 5]
```diff
@@ -308,23 +310,22 @@ def render_revenue_history
 
       def render_abilities(abilities)
         attribute_lines = []
+        props = { style: { display: 'inline-block' } }
```
Comment (tobymao) [inline assets/app/view/game/corporation.rb:313]: i think we need a bit of margin / padding for the abilities

File: lib/engine/config/game/g_1846.rb [hunk 6]
```diff
@@ -399,6 +399,7 @@ module G1846
       "abilities": [
         {
           "type": "token",
+          "name": "Reserved $40/$60 Ft. Wayne token",
```
Comment (tobymao) [inline lib/engine/config/game/g_1846.rb:402]: i think we shouldn't use name, it should be description

can you change the ability.name to be ability.description?

File: lib/engine/step/g_18_al/token.rb [hunk 7]
```diff
@@ -10,7 +10,7 @@ def place_token(entity, city, token, teleport: false)
           super
 
           entity.abilities(:assign_hexes) do |ability|
-            next unless ability.name == 'Historical objective'
+            next unless ability.name.start_with == 'Historical objective'
```
Comment (tobymao) [inline lib/engine/step/g_18_al/token.rb:13]: this is not a method?

it should be name.start_with?('Historical objective')

**Comment by yzemaze** [comment]
> 1846 updated version:

Left-aligned looks really wonky, even out of place (quite literally ;) ). If most abilities will be on companies’/minors’ cards center-aligned will fit way better. After toby is done with the specials rewrite I’ll look into headers of company tables to fit the centered style better.

File: assets/app/view/game/corporation.rb [hunk 8]
```diff
@@ -308,23 +310,22 @@ def render_revenue_history
...
+          attribute_lines << h('div.nowrap', props, ability.name)
         end
 
         table_props = {
           style: {
-            padding: '0 0.5rem',
-            grid: 'auto / 3fr 2fr',
-            gap: '0 0.2rem',
+            padding: '0.5rem',
+            grid: 'auto',
```
Comment (yzemaze) [inline assets/app/view/game/corporation.rb:322]: grid & gap are obsolete if you’re just using lines

File: lib/engine/config/game/g_1846.rb (hunk 1 above)
Comment (perwestling) [reply lib/engine/config/game/g_1846.rb]: Fixed

File: lib/engine/config/game/g_1846.rb (hunk 2 above)
Comment (perwestling) [reply lib/engine/config/game/g_1846.rb]: Fixed

File: assets/app/view/game/corporation.rb (hunk 3 above)
Comment (perwestling) [reply assets/app/view/game/corporation.rb:52]: Fixed

File: assets/app/view/game/corporation.rb (hunk 4 above)
Comment (perwestling) [reply assets/app/view/game/corporation.rb:54]: Fixed

File: lib/engine/config/game/g_1846.rb (hunk 6 above)
Comment (perwestling) [reply lib/engine/config/game/g_1846.rb:402]: ok, changed

File: lib/engine/step/g_18_al/token.rb (hunk 7 above)
Comment (perwestling) [reply lib/engine/step/g_18_al/token.rb:13]: I removed the name/description from the configuration and check for hexes size instead. The other asssign_hexes ability in 18AL has more than 1 hexes to choose from.

File: assets/app/view/game/corporation.rb (hunk 8 above)
Comment (perwestling) [reply assets/app/view/game/corporation.rb:322]: Fixed

File: assets/app/view/game/corporation.rb (hunk 4 above)
Comment (perwestling) [reply assets/app/view/game/corporation.rb:54]: Fixed

File: assets/app/view/game/corporation.rb (hunk 3 above)
Comment (perwestling) [reply assets/app/view/game/corporation.rb:52]: Fixed

File: lib/engine/config/game/g_1846.rb (hunk 6 above)
Comment (perwestling) [reply lib/engine/config/game/g_1846.rb:402]: Ok, fixed

File: lib/engine/step/g_18_al/token.rb (hunk 7 above)
Comment (perwestling) [reply lib/engine/step/g_18_al/token.rb:13]: Yeah, was a bug. But I changed this check to checking for hexes.length instead, so I could remove name/description from the config file.

File: assets/app/view/game/corporation.rb (hunk 8 above)
Comment (perwestling) [reply assets/app/view/game/corporation.rb:322]: Removed now. And put in centralized as you had in your commit.

File: assets/app/view/game/corporation.rb (hunk 5 above)
Comment (perwestling) [reply assets/app/view/game/corporation.rb:313]: I tried to add padding here but it just moved the ability lines a bit to the right, and kept Ability header as is, which looked fishy.

Any suggestion what should be done?

**Comment by perwestling** [comment]
After the latest push 1846 looks like this:
Screenshot from 2020-07-25 13-03-23

And when ability and company appear both on the corporation charter it looks like this:
Screenshot from 2020-07-25 13-27-22

**Pull request merged**